File: scripts/fetch_aphis_data.py
"""
Fetch HPAI wild bird detections CSV from USDA APHIS and convert to JSON.
Used by GitHub Actions to keep dashboard data current.
"""
import csv
import json
import sys
import urllib.request

APHIS_CSV_URL = "https://www.aphis.usda.gov/sites/default/files/hpai-wild-birds.csv"
OUTPUT_FILE = "data/hpai-wild-birds.json"

# Map CSV column names to the property names the dashboard expects
COLUMN_MAP = {
    "State": "State",
    "County": "County",
    "Collection Date": "Collection_Date",
    "Date Detected": "Date_Detected",
    "HPAI Strain": "HPAI_Strain",
    "Bird Species": "Bird_species",
    "WOAH Classification": "WOAH_Classification",
    "Sampling Method": "Sampling_Method",
    "Submitting Agency": "Submitting_Agency",
}


def fetch_and_convert():
    print(f"Fetching CSV from {APHIS_CSV_URL} ...")
    req = urllib.request.Request(APHIS_CSV_URL, headers={"User-Agent": "HPAI-Dashboard/1.0"})
    with urllib.request.urlopen(req) as resp:
        raw = resp.read().decode("utf-8-sig")

    reader = csv.DictReader(raw.splitlines())
    records = []
    for row in reader:
        record = {}
        for csv_col, js_key in COLUMN_MAP.items():
            record[js_key] = row.get(csv_col, "").strip()
        records.append(record)

    print(f"Parsed {len(records):,} records")

    with open(OUTPUT_FILE, "w", encoding="utf-8") as f:
        json.dump(records, f, separators=(",", ":"))

    print(f"Wrote {OUTPUT_FILE}")


if __name__ == "__main__":
    fetch_and_convert()

File: scripts/fetch_aphis_data.js
const https = require('https');
const fs = require('fs');
const path = require('path');

const url = 'https://www.aphis.usda.gov/sites/default/files/hpai-wild-birds.csv';
const outFile = path.join(__dirname, '..', 'data', 'hpai-wild-birds.json');

const COLUMN_MAP = {
  'State': 'State',
  'County': 'County',
  'Collection Date': 'Collection_Date',
  'Date Detected': 'Date_Detected',
  'HPAI Strain': 'HPAI_Strain',
  'Bird Species': 'Bird_species',
  'WOAH Classification': 'WOAH_Classification',
  'Sampling Method': 'Sampling_Method',
  'Submitting Agency': 'Submitting_Agency'
};

function parseCSVLine(line) {
  const result = [];
  let current = '';
  let inQuotes = false;
  for (let i = 0; i < line.length; i++) {
    const ch = line[i];
    if (ch === '"') {
      inQuotes = !inQuotes;
    } else if (ch === ',' && !inQuotes) {
      result.push(current.trim());
      current = '';
    } else {
      current += ch;
    }
  }
  result.push(current.trim());
  return result;
}

https.get(url, res => {
  let data = '';
  res.on('data', chunk => { data += chunk; });
  res.on('end', () => {
    // Remove BOM if present
    if (data.charCodeAt(0) === 0xFEFF) data = data.slice(1);

    const lines = data.split('\n').filter(l => l.trim());
    const headers = parseCSVLine(lines[0]);
    const records = [];

    for (let i = 1; i < lines.length; i++) {
      const vals = parseCSVLine(lines[i]);
      if (vals.length < headers.length) continue;
      const rec = {};
      headers.forEach((h, j) => {
        const key = COLUMN_MAP[h] || h;
        rec[key] = vals[j] || '';
      });
      records.push(rec);
    }

    fs.writeFileSync(outFile, JSON.stringify(records));
    console.log('Wrote ' + records.length.toLocaleString() + ' records to ' + outFile);
  });
}).on('error', err => {
  console.error('Fetch failed:', err.message);
  process.exit(1);
});
